R/merge_sce_list.R

```diff
@@ -0,0 +1,199 @@
+#'  in the main experiment of the final merged SCE object.
+#' @param preserve_rowdata_cols A vector of column names that appear in originating
+#'  SingleCellExperiment objects' rowData slots which should not be renamed with
+#'  the given SingleCellExperiment object name. These are generally columns which are
+#'  not specific to the given library's preparation or statistics. For example,
+#'  such an array might contain items like "Gene", "ensembl_ids", "gene_symbol", etc.
+#'
+#' @return A SingleCellExperiment object containing all SingleCellExperiment objects
+#'   present in the inputted list
+#' @export
+#'
+#' @import SingleCellExperiment
+merge_sce_list <- function(sce_list = list(),
+                           batch_column = "library_id",
+                           barcode_column = "barcode",
+                           retain_coldata_cols = c("sum",
+                                                   "detected",
+                                                   "total",
+                                                   "subsets_mito_sum",
+                                                   "subsets_mito_detected",
+                                                   "subsets_mito_percent",
+                                                   "miQC_pass",
+                                                   "prob_compromised"),
+                           preserve_rowdata_cols = NULL) {
+
+  # Check `sce_list`----------------------
+  if (is.null(names(sce_list))) {
+    stop("Individual SingleCellExperiment objects in `sce_list` must be named.")
+  }
+  if (length(sce_list) < 2) {
+    warning("There are fewer than two SCE objects in the provided `sce_list` so there is nothing to merge.")
+    # Early return:
+    return(sce_list)
+  }
+
+  # Check `retain_coldata_cols` ----------------
+  if (length(retain_coldata_cols) == 0) {
+    warning("All colData will be removed from the the merged SCE.
+            Please check that `retain_coldata_cols` was correctly specified.")
+  }
+
+  # Subset SCEs to shared features and ensure appropriate naming ------------------
+
+  # First, obtain intersection among all SCE objects
+  shared_features <- sce_list |>
+    purrr::map(rownames) |>
+    purrr::reduce(intersect)
+
+  if (length(shared_features) == 0) {
+    stop("There are no shared features among provided SCE objects.
+         They cannot be merged.")
+  }
+
+  # Second, determine all the column names that are present in any SCE so it can
+  #  be created in any missing SCEs with `NA` values
+  all_colnames <- purrr::map(sce_list, ~names(colData(.))) |>
+    unlist() |>
+    unname() |>
+    unique()
+
+  # Check that the `retain_coldata_cols` are present in at least one SCE, and
+  #  error if the column exists nowhere.
+  if (!(any(retain_coldata_cols %in% all_colnames))) {
+    stop("Error: The provided `retain_coldata_cols` are not present in any SCEs.")
+  }
```

Comment:
The comment makes it sound like you are checking that _all_ retain columns are found in at least one SCE, but the check is for at least one of the columns. Is that correct? I think it makes sense the way the comment is written, so `all()` should work here.